File: fly/src/main/java/com/tiamosu/fly/integration/gson/data/BigDecimalTypeAdapter.java
package com.tiamosu.fly.integration.gson.data;

import androidx.annotation.NonNull;

import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;
import java.math.BigDecimal;

/**
 * author : Android 轮子哥
 * github : https://github.com/getActivity/GsonFactory
 * time   : 2021/01/01
 * desc   : BigDecimal 类型解析适配器，参考：{@link com.google.gson.internal.bind.TypeAdapters#BIG_DECIMAL}
 */
public class BigDecimalTypeAdapter extends TypeAdapter<BigDecimal> {

    @Override
    public BigDecimal read(JsonReader in) throws IOException {
        switch (in.peek()) {
            case NUMBER:
            case STRING: {
                final String result = in.nextString();
                if (result == null || "".equals(result)) {
                    return new BigDecimal(0);
                }
                return new BigDecimal(result);
            }
            case NULL: {
                in.nextNull();
                return new BigDecimal(0);
            }
            default: {
                in.skipValue();
                return new BigDecimal(0);
            }
        }
    }

    @Override
    public void write(@NonNull JsonWriter out, BigDecimal value) throws IOException {
        out.value(value);
    }
}

File: fly/src/main/java/com/tiamosu/fly/integration/gson/data/BigDecimalTypeAdapter.kt
package com.tiamosu.fly.integration.gson.data

import com.google.gson.TypeAdapter
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonToken
import com.google.gson.stream.JsonWriter
import java.io.IOException
import java.math.BigDecimal

/**
 * @author tiamosu
 * @date 2021/2/1.
 */
internal class BigDecimalTypeAdapter : TypeAdapter<BigDecimal>() {

    @Throws(IOException::class)
    override fun read(`in`: JsonReader?): BigDecimal {
        if (`in` == null) {
            return BigDecimal(0)
        }
        return when (`in`.peek()) {
            JsonToken.NUMBER, JsonToken.STRING -> {
                val result = `in`.nextString()
                if (result == null || "" == result) BigDecimal(0) else BigDecimal(result)
            }
            JsonToken.NULL -> {
                `in`.nextNull()
                BigDecimal(0)
            }
            else -> {
                `in`.skipValue()
                BigDecimal(0)
            }
        }
    }

    @Throws(IOException::class)
    override fun write(out: JsonWriter, value: BigDecimal?) {
        out.value(value ?: BigDecimal(0))
    }
}